### ISSUE TITLE
onProductPurchased & consumePurchase() not getting called when the purchase is done. 

### ISSUE THREAD

**Issue by aakashvats2910** [anjlab/android-inapp-billing-v3#450]
Whenever i make a purchase the ```onProductPurchased``` method is not called right away. It gets called only after when the payment is proceeded second time and then gives productId and details.

Same is happening with the ```bp.consumePurchase()```

**I think its a bug and not a feature, so please help me right on this, need help ASAP**

**My logcat is this :** (In this you can find my debug events starting with  "()()()()"
2020-07-21 20:03:27.259 7453-7481/org.kashisol.convertit D/mali_winsys: EGLint new_window_surface(egl_winsys_display *, void *, EGLSurface, EGLConfig, egl_winsys_surface **, EGLBoolean) returns 0x3000
2020-07-21 20:03:27.263 7453-7453/org.kashisol.convertit I/System.out: **()()()() Billing Initialized**
2020-07-21 20:03:27.278 7453-7453/org.kashisol.convertit I/Choreographer: Skipped 4 frames!  The application may be doing too much work on its main thread.
2020-07-21 20:03:27.309 7453-7453/org.kashisol.convertit W/RenderThread: type=1400 audit(0.0:159093): avc: denied { search } for name="/" dev="mmcblk0p40" ino=2 scontext=u:r:untrusted_app:s0:c243,c258,c512,c768 tcontext=u:object_r:unlabeled:s0 tclass=dir permissive=0
2020-07-21 20:03:27.314 7453-7481/org.kashisol.convertit E/ion: ioctl c0044901 failed with code -1: Invalid argument
2020-07-21 20:03:27.419 7453-7453/org.kashisol.convertit D/WindowClient: Add to mViews: com.color.internal.widget.ColorToastLayout{ca3656b V.E...... ......I. 0,0-0,0 #c0204f9 oppo:id/color_toast_layout}, this = android.view.WindowManagerGlobal@2f14114
2020-07-21 20:03:27.421 7453-7453/org.kashisol.convertit D/ViewRootImpl[Toast]: hardware acceleration = true , fakeHwAccelerated = false, sRendererDisabled = false, forceHwAccelerated = false, sSystemRendererDisabled = false
2020-07-21 20:03:27.426 7453-7453/org.kashisol.convertit I/InputTransport: Create ARC handle: 0x7301db5d60
2020-07-21 20:03:27.427 7453-7453/org.kashisol.convertit I/Choreographer: Skipped 8 frames!  The application may be doing too much work on its main thread.
2020-07-21 20:03:27.453 7453-7453/org.kashisol.convertit D/Surface: Surface::allocateBuffers(this=0x72d907c000)
2020-07-21 20:03:27.453 7453-7481/org.kashisol.convertit D/Surface: Surface::connect(this=0x72d907c000,api=1)
2020-07-21 20:03:27.455 7453-7481/org.kashisol.convertit D/mali_winsys: EGLint new_window_surface(egl_winsys_display *, void *, EGLSurface, EGLConfig, egl_winsys_surface **, EGLBoolean) returns 0x3000
2020-07-21 20:03:27.492 7453-7453/org.kashisol.convertit I/Choreographer: Skipped 1 frames!  The application may be doing too much work on its main thread.
2020-07-21 20:03:29.282 7453-7481/org.kashisol.convertit D/Surface: Surface::disconnect(this=0x72d907c000,api=1)
2020-07-21 20:03:29.284 7453-7453/org.kashisol.convertit D/View: [Warning] assignParent to null: this = com.color.internal.widget.ColorToastLayout{ca3656b V.E...... ......ID 0,0-438,105 #c0204f9 oppo:id/color_toast_layout}
2020-07-21 20:03:29.284 7453-7453/org.kashisol.convertit I/InputTransport: Destroy ARC handle: 0x7301db5d60
2020-07-21 20:03:32.471 7453-7453/org.kashisol.convertit I/System.out: **()()()() BUTTON CLICKED FOR PURCHASE**
2020-07-21 20:03:32.498 7453-7453/org.kashisol.convertit I/Choreographer: Skipped 2 frames!  The application may be doing too much work on its main thread.
2020-07-21 20:03:40.908 7453-7453/org.kashisol.convertit V/PhoneWindow: DecorView setVisiblity: visibility = 0, Parent = android.view.ViewRootImpl@c04fdbd, this = DecorView@bd7af67[MainActivity]
2020-07-21 20:03:40.908 7453-7453/org.kashisol.convertit I/Choreographer: Skipped 1 frames!  The application may be doing too much work on its main thread.
2020-07-21 20:03:47.225 7453-7453/org.kashisol.convertit I/System.out: **()()()() BUTTON CLICKED FOR PURCHASE**
2020-07-21 20:03:47.273 7453-7453/org.kashisol.convertit I/System.out: **()()()() ProductPurchased**:{"orderId":"GPA.3369-6279-9003-73554","packageName":"org.kashisol.convertit","productId":"one_fifty_deal","purchaseTime":1595342017549,"purchaseState":0,"developerPayload":"inapp:one_fifty_deal:2bd64182-b7a2-4796-ae4d-a8034a0793bc","purchaseToken":"aagblhcoadeogamhhcfhkadf.AO-J1OyWfHQk6nsgLo5guUV76i6-FJyG50w85J2m7EPyQCE8Ryj6qy3p18yX-8UcThBJ2vMPIjg5lgfd0qWu613QYIlhD_ld3gX83nZgC5HKsb4gAiGKmucSTUhp5Y1lDNsPx4ojG35d"}
2020-07-21 20:03:48.316 7453-7453/org.kashisol.convertit D/iabv3: Successfully consumed one_fifty_deal purchase.
2020-07-21 20:03:48.318 7453-7453/org.kashisol.convertit I/Choreographer: Skipped 65 frames!  The application may be doing too much work on its main thread.

**NOTE :-** After the button was click for first time the purchase was done, but details were shown only when clicked 2nd time.

**Comment by dayron9110** [comment]
Any updates?

**Comment by matrixreal** [comment]
same here, any update?

**Comment by aakashvats2910** [comment]
@matrixreal @dayron9110 no updates from the developer till now 👎

**Comment by matrixreal** [comment]
> @matrixreal @dayron9110 no updates from the developer till now 👎

try this on oncreate :
bp = BillingProcessor XXXXXXXXXXXXXXXXXXXXXX
        bp.initialize()
        bp.getPurchaseTransactionDetails(getString("product_id")
        img.setOnClickListener{
            bp.purchase(this, getString("product_id")
        }
the idea is to call the bp.getPurchaseTransactionDetails(getString("product_id") after initialisation and before the purchase

**Comment by nageamol** [comment]
Any Update?? same issue.

**Comment by aakashvats2910** [comment]
> Any Update?? same issue.

Didn't check, I had moved to the official API by google.

**Comment by nageamol** [comment]
Yes i did same and Google's official lib is simpler than this. Thanks for update
